system/jlib/jfile.cpp

```diff
@@ -2027,18 +2042,24 @@ void CFileIO::close()
 #ifdef CFILEIOTRACE
         DBGLOG("CFileIO::close(%d), extraFlags = %d", tmpHandle, extraFlags);
 #endif
-        if (extraFlags & IFEnocache)
+        if (extraFlags & (IFEnocache | IFEsync))
         {
             if (openmode != IFOread)
             {
+                if (extraFlags & IFEsync)
+                {
 #ifdef F_FULLFSYNC
-                fcntl(tmpHandle, F_FULLFSYNC);
+                    fcntl(tmpHandle, F_FULLFSYNC);
 #else
-                fdatasync(tmpHandle);
+                    fdatasync(tmpHandle);
 #endif
+                }
+                else
+                    sync_file_region(tmpHandle, 0, 0, false);
```

Comment:
prior to this change, CFileIO::close with IFEnocache would call fdatasync and posix_fadvise(POSIX_FADV_DONTNEED),
and CFileIO::write would call sync_file_region and posix_fadvise(POSIX_FADV_DONTNEED).

After this change, both will call sync_file_region and posix_fadvise(POSIX_FADV_DONTNEED).

Are both needed to indicate pages can be discarded?

Reply:
Yes.  sync_file_region() requests dirty blocks to be commited - async or sync depending on the flags provided.
posix_fadvise(DONTNEED) is to tell the kernel to drop clean (already commited) blocks from the page-cache.
You need both, you cannot drop dirty blocks from page-cache.